Mentions in comments backend changes

cmrd-senya opened the pull request:
This is an early preliminary PR with backend-changes to support mentions in comments (#1851).

It introduces changes to the DB schema making Mention model polymorphic and allowing a mention to belong to a comment. It also creates module Diaspora::MentionsContainer which incapsulates the logic of an object that supports keeping mentions. Now it includes logic I picked from the StatusMessage model.

ATM the mentions in comments are supported to the extent of placing user profile links instead of mention code in the mobile version (it is so because the mobile version doesn't require javascript code changes).

Changes summary:
- Make Mention model polymorphic;
- Introduce Diaspora::MentionsContainer which incapsulates the logic of an object that supports keeping mentions;
- We don't filter mentions anymore, instead we send notification/show in streams only that mentions, which are visible to the recipient;
- Create Notifications::Mentioned module which incapsulates common logic of mentioned notification;
- Add Notifications::MentionedInComment and move old Notifications::Mentioned to Notifications::MentionedInPost;
- Update the UI code to render new type of notification;
- Hack comments notifications to avoid having double notification when mentioned in a comment;

Drive-by refactorings:
- Move Notification.types out of the model to the controller;
- Move `StatusMessage#mentioned_people_names` out of the model. Create a `PersonPresenter.people_names` method instead;
- DRY UserController to use UserPreference::VALID_EMAIL_TYPES in the `user_params` method;

TODO:
- [x] Implement notifications for mentioning in comments
- [x] Implement restrictions of mentioning some kind of people (see [comment](https://github.com/diaspora/diaspora/issues/1851#issuecomment-36557842))
- [x] Ensure comment is federated to new discussion participants
- [x] Ensure mentions in comment are filtered on federation receive
- [x] ~~Ensure the filtered out mentions are replaced with links properly~~
- [x] Don't send redundant second notification when mentioned in comment.
- [x] Ensure email notifications work properly
- ~~Implement frontend javascript changes for comments publisher~~

SuperTux88 commented on app/models/person.rb:
```diff
@@ -172,6 +172,10 @@ def self.search_order
     }.call
   end
 
+  def self.people_names(people)
```
```
## TODO ----
# don't put presentation logic in the model!
```

Also, this is only used once, why don't you just move `.map(&:name).join(", ")` to where its used? Maybe the controller is also not the best place, but it's better than split this and add it to the model.

cmrd-senya replied:
Well, I like to have it as a separate method, if possible. I don't know of a better place for that though. Any ideas?

SuperTux88 commented on db/schema.rb:
```diff
@@ -211,13 +211,14 @@
   end
 
-  add_index "mentions", ["person_id", "post_id"], name: "index_mentions_on_person_id_and_post_id", unique: true, using: :btree
+  add_index "mentions", ["mentions_container_id"], name: "index_mentions_on_mentions_container_id", using: :btree
```
you should also add `mentions_container_type` to this index, because it is never used without the type.

SuperTux88 commented on app/models/mention.rb:
```diff
@@ -5,16 +5,22 @@
 
-  belongs_to :post
+  belongs_to :mentions_container, polymorphic: :true
+  belongs_to :post, ->(mention) { (mention.mentions_container_type == "Post") ? where("1 = 1") : none },
```
WTF? what does this? and why?

cmrd-senya replied:
This allow to write
```
Mention.where(mentions_container_id: @post.id, mentions_contrainer_type: "Post")
```
as
```
Mention.where(post: @post)
```

SuperTux88 replied:
ah, fancy stuff.

But when you do
```ruby
Mention.where(post: Post.find(post_id))
```
it needs two querys and I don't know why this should be better than
```ruby
Mention.where(mentions_container_id: post_id, mentions_contrainer_type: "Post")
```

cmrd-senya replied:
I picked the idea from the second answer [here](http://stackoverflow.com/questions/21850995/association-for-polymorphic-belongs-to-of-a-particular-type). I don't like the self-reference approach from the first answer, it looks less-obvious for me.

cmrd-senya replied:
Ok, I'll check it once again then. One of my purposes for that was to escape rewriting all the code which references to Mention through post.

cmrd-senya commented:
Ok, I have completed backend support for mentions in comments here. Front end changes are completely separate thing and I suppose they can be implemented&reviewed separetely. This PR can be tested using the markdown syntax of mention directly: `@{alice; alice@example.tld}`.

Can be reviewed now.

jhass commented on app/helpers/notifications_helper.rb:
```diff
@@ -4,40 +4,46 @@ module NotificationsHelper
+      post_link: link_to(post_page_title(post), post_path(post)).html_safe
+    }.tap {|opts|
+      if note.linked_object.instance_of?(Comment)
+        opts.merge!(comment_path: "#{post_path(post)}##{note.linked_object.guid}".html_safe)
```
IIRC path helpers take an `anchor` option, `post_path(post, anchor: note.linked_object.guid)`.

jhass commented on lib/diaspora/mentions_container.rb:
```diff
@@ -0,0 +1,37 @@
+        after_create :create_mentions
+        has_many :mentions, as: :mentions_container, dependent: :destroy
+      end
+    end
```
Well, we're inside Rails, so it might be more idiomatic to use `AS::Concern` here.

```rb
  extend ActiveSupport::Concern
  included do
    ....
  end
````

jhass commented on app/models/comment.rb:
```diff
@@ -46,14 +48,17 @@ class Comment < ActiveRecord::Base
+    return @people_allowed_to_be_mentioned unless @people_allowed_to_be_mentioned.nil?
+    contacts = author.owner.try(:contact_people) if parent.public?
+
+    @people_allowed_to_be_mentioned = parent.try(:participants).to_a.concat(contacts.to_a).push(parent.author).uniq
```
Can you try whether `AR::Relation is splattable? `[*parent.try(:participants), *contacts, parent.author].uniq` Though even with explicit `to_a`s that might be a little bit more readable :)

cmrd-senya replied:
> I'm not quite sure I follow why we know parent.author always succeeds but parent.participants may not

We have a [spec](https://github.com/diaspora/diaspora/blob/develop/spec/models/photo_spec.rb#L209) for photos commenting, and photos don't have participants. This spec fails when I do `parent.participants` with no `try`. Am I right that we don't have separate photos comments allowed now?

Flaburgan replied:
>Am I right that we don't have separate photos comments allowed now?

At the moment, the interface doesn't allow users to comment on a specific photo, they can only comment on the post which contains the photo. But I guess this can change in the future.

cmrd-senya replied:
So I guess we should leave it as a resposibility of Photo to implement `participants` and stay with `try` or maybe stub this method with `[]` for now. Which is better?

jhass replied:
I'm not sure, I think I would get rid of the spec for now. I don't see reintroducing photo specific comments coming any time soon.

cmrd-senya replied:
> Can you try whether AR::Relation is splattable?
Seems so.

> I think I would get rid of the spec for now.
Ok, then I'll remove the spec if nobody minds.

jhass commented:
Sorry I don't remember, since we allow to bring new people into the conversation by allowing the commenter to mention anyone of their contacts, did we decide how to federate the comment in case the mentioned person never got the post initially?

cmrd-senya commented:
I walked quickly through the discussions and as far as I can see there was no decision. We allow to bring new people into the conversation for the public posts only. In this case we can add a mentioned person to the list of the comments subscribers and the comment will be federated to a new pod. After that, the new receiver pod may fetch the post if it's yet unknown. AFAIK, with the new federation it is nothing that we can't do.

Also I'm a bit concerned with the way notifications work in this PR. In cases when we mention a participant, the participant received two notifications (e.g. AlsoCommented and MentionedInComment). That's not too elegant. I'll seek for a way to mute CommentOnPost and AlsoCommented in case when we send MentionedInComment to a person.

Since these two issues are unresolved in the PR I'll set the status to [WIP] again.

Commit:
Remove the photo comment spec

This functionality is not something we support and it doesn't look
it will be reintroduced soon.

see discussion https://github.com/diaspora/diaspora/pull/6818#discussion_r70227392

SuperTux88 commented on app/models/notifications/mentioned_in_comment.rb:
```diff
@@ -0,0 +1,27 @@
+      # mentioned in comment.
+      #
+      # I left intact the Notifications::Mentioned.notify and didn't investigate the message dispatcher
+      # but here I add parent_owner_id to the list so the filtering passes well.
```
This doesn't work, because private posts are received multiple times (to create the ShareVisibilities), so the mention-notification is only sent, if the user is in the `recipient_ids` (the user who received the visibility). But comments are only received once (all others are ignored), because all visibilities are already there. See `AlsoCommented`.

cmrd-senya replied:
Hmm, I just tried to remove this method and everything seems to work fine. I made a mistake, thanks for pointing!

SuperTux88 replied:
No, that will not work either. If you use the parent-method, it will work for public posts, but for private-posts it will only send a notification, if the mentioned person is the first who receives it.

I'm not even sure if we should extend from `Mentioned`, because the `notify`-method would be simpler (than the parent) if you completely write the logic, instead of trying to use some parent logic.

cmrd-senya replied:
> No, that will not work either. If you use the parent-method, it will work for public posts, but for private-posts it will only send a notification, if the mentioned person is the first who receives it.

Ah, ok, that means my tests don't cover that :)

> I'm not even sure if we should extend from Mentioned, because the notify-method would be simpler (than the parent) if you completely write the logic, instead of trying to use some parent logic.

Yes, that what I probably should do.

cmrd-senya commented:
@SuperTux88, I ended up with keeping inheritance. Please check 048d6b0777ed6065a9c2df9cb821b4a5e689e4d9 and tell me if you have any objections on the approach.

SuperTux88 commented on app/models/notifications/mentioned.rb:
```diff
@@ -13,19 +13,25 @@ def deleted_translation_key
       end
     end
+
+    def self.get_mentions(mentionable, recipient_user_ids)
```
I think we shouldn't use `get_`, rubocop has a cop for this, but only for instance methods.

SuperTux88 commented on lib/diaspora/mentions_container.rb:
```diff
@@ -0,0 +1,41 @@
+      has_many :mentions, as: :mentions_container, dependent: :destroy
+    end
+
+    def local_mentions
```
I would add a scope `local` to the `Mention` model, so you can use `mentionable.mentions.local`

SuperTux88 commented on app/models/notifications/mentioned.rb:
```diff
@@ -13,19 +13,25 @@ def deleted_translation_key
     end
+
+    def self.get_mentions(mentionable, recipient_user_ids)
+      relation = mentionable.local_mentions.where.not(person: mentionable.author)
```
this is the same line as in the `MentionedInComment`

SuperTux88 commented on app/models/notifications/mentioned.rb:
```diff
@@ -13,19 +13,25 @@ def deleted_translation_key
+      if mentionable.public
+        relation
+      else
+        relation.where(person_id: recipient_user_ids.map {|id| User.find(id).person_id })
```
`Person.where(owner_id: recipient_user_ids).ids`

SuperTux88 commented on app/models/comment.rb:
```diff
@@ -64,6 +75,12 @@ def initialize(person, target, text)
 
+    def build(*args)
+      super.tap do |comment|
+        comment.filter_mentions(comment.people_allowed_to_be_mentioned)
```
We only filter this on send? and not on receive? Or do I miss something? So what happens, if we receive a comment with a mention (maybe from another network that doesn't have this outgoing filter), and the mentioned user can not see the post? does he receive a notification?

cmrd-senya replied:
It doesn't seem to be filtered properly at the receive via federation, but that probably also applies to mentions in StatusMessage since the #6902 merge (or earlier). Gonna write some tests for that.

SuperTux88 replied:
Yes, maybe status messages are also not properly filtered on receive. It was just an idea, that we may should also filter on receive, so we don't send notifications, if the user can't see the post. 😃

cmrd-senya replied:
So [looks like](https://travis-ci.org/diaspora/diaspora/jobs/147856327#L653) mentioning in status message and post via federation may bypass some visibility restrictions.

SuperTux88 replied:
I discovered that this incoming filter is a bit more complex than I thought first.

For **public** stuff it's easy: don't filter anything, the sender can mention anyone. You can't know the senders aspects, so you don't know if he is allowed to mention someone.

For **private comments** it's easy too: you can only mention other participants, every pod can validate and filter that.

For **private posts** it's a bit more complex: We can not know if the mentioned person can see the post, when we receive it (that's what I not considered before). So we can not filter anything here and need to trust the sender. But my notification logic already handles the notifications correctly, so the person only receives a notification, when he receives the visibility. So maybe we need to ensure that the person can not see the post on his stream, unless he also receives the visibility, instead of filter the mention?

So what you built now is, that you filter all mentions on private posts on receive, which makes mention impossible on private posts.

cmrd-senya replied:
> So what you built now is, that you filter all mentions on private posts on receive, which makes mention impossible on private posts.

Not really. Have a look on [this](https://github.com/diaspora/diaspora/pull/6818/files#diff-2bcec3a9d410c74617f05ba3d8408810R156) test -- it passes well, that means that mentions work on private receive. Actually, what I built now is no filtering at all (in the federation case). It rather creates mentions only for people who are allowed to be mentioned. However it doesn't work well right now in the case of multiple receive again.

The question is when do we create Mention objects for post/comment. Mention objects was originally created only when mention is allowed. That's why filtering was used - to avoid creation of Mention objects. If we change the logic of stream rendering so it hides posts when person is not allowed to see them irrespective of Mention object presence, then we could create Mention objects for all mentioned people immediately after receive. I guess this is what you suggest.

After some thoughts, I agree that this is more optimal way to implement visibility restriction for mentioning. This way we won't need any filtering at all and get more straightforward design.

SuperTux88 replied:
> Not really. Have a look on this test -- it passes well, that means that mentions work on private receive.

That test is not really useful:
1. you don't know who can see this post on receive, because you only receive the first visibility.
2. my local test (where I saw that it doesn't work with private) was: podA  has your code, user@podB mentions user@podC (which is also a receiver of this post), but the mention was filtered (because podA doesn't know about that).

You can filter local users and recreate filtered mentions on second receive, if the second user can see it now, but I doubt that this is practicable.

So that all is no real solution for the problem. I think the most reasonable solution is to filter the streams, so you can only view posts, that you are allowed to see. We should still filter on outgoing stuff (because old posts would otherwise show the posts to users who shouldn't see it), but incoming private posts are 
impossible to filter, because you don't know who can see it.

But my original intention with this topic (filter on receive) was, to ensure that a user don't receives a notification for a mention in a comment, if he can't see the post. If this is done on notification-level (like for notifications for mentions in posts, where this already works), I'm totally fine with it.

SuperTux88 commented on app/models/status_message.rb:
```diff
@@ -121,6 +89,13 @@ def post_location
     }
   end
 
+  def people_allowed_to_be_mentioned
```
where is this method called?

cmrd-senya replied:
This method was replaced with [this line](https://github.com/diaspora/diaspora/pull/6818/files#diff-1b42e0e2c5d972f2a8ef8ea55b6b92dfR23) after rebase on 46cbc6e52a9d4a03dff8228ce11ac4876d247a15, so it's not used. Thanks for noticing.

SuperTux88 commented on app/models/notifications/mentioned.rb:
```diff
@@ -13,19 +13,26 @@ def deleted_translation_key
       end
     end
+
+    def self.filter_mentions(mentions, mentionable, recipient_user_ids)
```
So the point where I don't like the inheritance here is, that the method on the parent class is the "special case" (filter on private posts), and the same method on the child class, should be the default case (just no filtering). But invert the inheritance (extend `Mentioned` from `MentionedInComment`) feels also wrong.

So maybe the solution is to add a `AbstractMentioned` with the default (no filtering) `filter_mentions`, `notify` and `linked_object` method, and then both of `Mentioned` and `MentionedInComment` extending from this, and add the translation_keys and override the `filter_mentions` in the `Mentioned` class.

Just some thoughts ...

jhass replied:
Do we need classic inheritance at all? Or stub definitions in the shared code? I'd say extract a module for the common code and just the common code and include it into both.

cmrd-senya replied:
Actually, I like classic inheritance here more. @SuperTux88, what do you think?

If I extract a module, where do I put it and how do I call it? For example if I make `Notification::MentionedHelper` and put it somewhere in `lib/` it's less convenient than to have a base class along with children in the same folder.

jhass replied:
If you can argue for the base class, which is never directly used directly to live in the same folder, so you can for the module. If you can argue for the base class to be called something with an artificial suffix, like `MentionedBase`, so you can for the module (Something like `Mentioned` and `PostMentioned`, `CommentMentioned` would deprive from doing that in either case btw.)

Ruby has no abstract base classes, it uses modules to share functionality.

jhass replied:
Btw a pattern that emerges in the Rails community is to have a `concerns` subfolder and in there modules (that extend `ActiveSupport::Concern`). I'm not yet sure whether I like that, but it's another option.

cmrd-senya replied:
Sounds convincing. So is everybody ok with putting the shared module with common code in `app/models/notification`? I think I can call it `MentionedShared` or `MentionedCommon`, or something like that.

I see that we already have some extension modules for `User` in our `app/models/user` folder which I can consider as an acceptable decision then.

jhass replied:
I think renaming `Mentioned` to `MentionedInPost` in order to free the name for the shared module is fair.

cmrd-senya replied:
Does it mean that I have to write a migration in order to change the `type` of notification in the DB  or can I just override the convention like this?
```ruby
class Notification::MentionedInPost < Notification
...
  def self.sti_name
      "Notification::Mentioned"
  end
...
end
```

jhass replied:
I think I would prefer a migration :)

SuperTux88 commented on spec/integration/mentioning_spec.rb:
```diff
@@ -23,73 +31,254 @@ def mention_stream_for(user)
+
+    allow_callbacks(%i(queue_private_receive receive_entity fetch_related_entity))
+
+    xml = generate_xml(entity, remote_user_on_pod_b, recipient)
```
You don't need this whole generate and parse xml here, just send the `entity` to the `:receive_entity` callback.

SuperTux88 commented on db/migrate/20160621151100_polymorphic_mentions.rb:
```diff
@@ -0,0 +1,12 @@
+    remove_index :mentions, %i(person_id post_id)
+    rename_column :mentions, :post_id, :mentions_container_id
+    add_column :mentions, :mentions_container_type, :string, default: "Post", null: false
+    add_index :mentions, %i(mentions_container_id mentions_container_type), name: "index_mentions_on_mc_id_and_mc_type"
```
the migration breaks on mysql:
```
Mysql2::Error: Specified key was too long; max key length is 767 bytes: CREATE  INDEX `index_mentions_on_mc_id_and_mc_type`  ON `mentions` (`mentions_container_id`, `mentions_container_type`)
```

cmrd-senya replied:
Works for me when I do
```
git checkout develop
RAILS_ENV=development FIXTURES_PATH=spec/fixtures bundle exec rake db:drop db:create db:schema:load db:fixtures:load
... posting post with mention ...
git checkout 1851-mentions-in-comments.new
bundle exec rake db:migrate
```
Does database content make difference here?
MySQL version is `mysqld  Ver 5.5.49-0ubuntu0.14.04.1 for debian-linux-gnu on x86_64 ((Ubuntu))`.

SuperTux88 replied:
When I try the same, I get the error: MariaDB 10.0.23

But I can create the index manually in the DB.

Anyway, why don't you just add the `length` parameter as in other migrations where indexes are added? WIth the `length` it works.

SuperTux88 commented on app/models/notifications/mentioned.rb:
```diff
@@ -13,19 +13,26 @@ def deleted_translation_key
     def self.notify(mentionable, recipient_user_ids)
       actor = mentionable.author
+      relevant_mentions = filter_mentions(
+        mentionable.find_or_create_mentions.local.where.not(person: mentionable.author),
```
why did you change this to `find_or_create_mentions`? The `mentions` should always be there at this point, if you create them in the `after_create`. The problem with `find_or_create_mentions` is, that this always parses the whole text again, if the post has NO mentions, because `mentions` would ALWAYS be empty for this posts, so it tries to create it. And `after_create` does this once. (same goes for comments without mentions)

cmrd-senya replied:
That is actually the same thing that we do know when we request `mentioned_people` for StatusMessage (see https://github.com/diaspora/diaspora/blob/develop/app/models/status_message.rb#L61).

However, we won't need this, if we introduce changes we discussed above in https://github.com/diaspora/diaspora/pull/6818#discussion-diff-72363123

SuperTux88 commented on spec/integration/mentioning_spec.rb:
```diff
@@ -1,13 +1,21 @@
 require "spec_helper"
 require "requests_helper"
+require "integration/federation/federation_helper"
```
Can you try to write this tests without the federation helper? When you don't use `remote_user_on_pod_b` and `remote_user_on_pod_c` (which add stubs on the callbacks which you don't need), you also don't need to use  `allow_callbacks` for the tests (no stubs == everything allowed with call original, which is what you want here).

Have a look at `remote_raphael`.

jhass commented on app/models/notifications/mentioned.rb:
```diff
@@ -1,38 +1,29 @@
-        recipient_user_ids
-      )
+    included do
+      def self.notify(mentionable, recipient_user_ids)
```
When using `AS::Concern`, instead of this, define a nested module called `ClassMethods` which have the class methods defined as instance methods.

SuperTux88 commented on db/migrate/20160621151100_polymorphic_mentions.rb:
```diff
@@ -0,0 +1,12 @@
+class PolymorphicMentions < ActiveRecord::Migration
+  def change
+    remove_index :mentions, %i(post_id)
```
This doesn't work for rollback: `remove_index is only reversible if given a :column option.`

SuperTux88 commented on app/services/status_message_creation_service.rb:
```diff
@@ -19,20 +19,9 @@ def create(params)
 
   def build_status_message(params)
     public = params[:public] || false
-    filter_mentions params
```
you should still filter on send, because otherwise old pods will display the post to people who shouldn't see it.

cmrd-senya replied:
fixed in https://github.com/diaspora/diaspora/pull/6818/commits/56dd62ceb39ee0f372eb549a71ac0df66a4f51c9

cmrd-senya commented:
Ok, this feels completed to me. Setting [WIP] off. It is ready to, hopefully, final review.

But I guess it won't be merged before the 0.6.0.0 release anyway? (Since there is no UI yet for mentioning in comment publisher, therefore it is not a complete feature).

SuperTux88 commented on app/models/comment.rb:
```diff
@@ -46,14 +48,23 @@ class Comment < ActiveRecord::Base
+
+  def subscribers
+    super.tap {|subscribers|
+      subscribers.concat(legitimate_mention_recipients) if parent.author.local?
```
This doesn't work for relayed comments on public posts, where the comment-author is allowed to mention his contacts. The parent author would filter this persons and don't deliver the comment to them if they are not also in the contacts of the parent-author or have already participated.

Maybe we should just allow all people to be mentioned on public comments (like on public posts)? We can decide to send no notification for the mentioned person, if he doesn't have the information, that the author is sharing with him, if that is a problem, but we also don't do that for posts.

cmrd-senya replied:
I don't see any other options either: we don't have means to restrict mentioned recipients if we want to include anyone except participants.

cmrd-senya replied:
The changes allowing everyone to be mentioned in public comments have been applied.

SuperTux88 replied:
When the (mentioned) post author is local and the (mentioning) comment author is remote, it sends two notification mails. That is because you add the local mentioned people (which already received it) again to the subscribers for relaying, so they receive it again.

SuperTux88 commented on config/locales/diaspora/en.yml:
```diff
@@ -634,9 +636,11 @@ en:
       one: "%{actors} mentioned you in a deleted post."
       other: "%{actors} mentioned you in a deleted post."
+    mentioned_in_comment_deleted:
+      one: "%{actors} mentioned you in a comment to a deleted post."
```
Maybe only the comment is deleted and not the whole post.

SuperTux88 commented on lib/diaspora/shareable.rb:
```diff
@@ -57,7 +65,7 @@ def owned_or_visible_by_user(user)
           visible_by_user(user).or(arel_table[:public].eq(true)
                                      .or(arel_table[:author_id].eq(user.person_id)))
-        ).select("DISTINCT #{table_name}.*")
+        ).distinct_wrapper
```
why? If `.distinct` doesn't work for postgres (I haven't tested that), just leave the `.select("DISTINCT #{table_name}.*")`, because the `.distinct` doesn't make something different for mysql.

cmrd-senya replied:
`.count` called on the relation doesn't work if `.select("DISTINCT #{table_name}.*")` used with mysql. `.distinct` fixes the thing for mysql, but breaks for postgres. You can reproduce this failure by reverting the change and running tests for `spec/models/status_message_spec.rb`.

SuperTux88 replied:
So, this is only for tests? I don't like changing the code (and maybe making it less readable) only for tests.

Also, if you change your test to
```ruby
expect(StatusMessage.where_person_is_mentioned(@bob).ids).to match_array([post1.id, post2.id])
```
it works with both databases without changing anything on the code. Also this would actually test if you see the correct posts, and not two other (maybe wrong) posts.

cmrd-senya replied:
TBH I haven't checked whether this is needed for anything in the production code. You are right, it doesn't worth changing anything for tests only. I'll verify that then.

SuperTux88 commented on spec/spec_helper.rb:
```diff
@@ -128,6 +128,11 @@ def client_assertion_with_nonexistent_client_id_path
   end
 
+  # Clear Warden environment
+  config.after(:each) do
```
Do we need a `config.after(:each)` for each one?

Also, if this fixes random rspec fails, then you should create its own small PR for it, which then probably could be merged faster than this big one.

Personally I think that also the `pry-rescue`-change should have its own PR, don't add unrelated stuff to a PR that is already big enough. If you need the changes here, you can rebase this PR on top of the other PRs.

SuperTux88 commented on app/models/notifications/also_commented.rb:
```diff
@@ -18,6 +18,7 @@ def self.notify(comment, _recipient_user_ids)
       recipient_ids = commentable.participants.local.where.not(id: [commentable.author_id, actor.id]).pluck(:owner_id)
 
       User.where(id: recipient_ids).find_each do |recipient|
+        next if Notifications::MentionedInComment.where(target: comment.mentions.where(person: recipient.person)).any?
```
`.exists?` instead of `.any?`

SuperTux88 commented:
Also, please cleanup your commits, squash some together (we don't need 6 "polishing" commits and fixups) and write meaningful messages for the rest.

SuperTux88 commented on lib/diaspora/mentions_container.rb:
```diff
@@ -0,0 +1,41 @@
+      if ppl == "all"
+        mentions
+      else
+        mentions.where(person: ppl)
```
I'm not sure about this, this generates a `WHERE IN (...)` query with potentially many people (when someone with many contacts shares a post with "all aspects").

cmrd-senya commented:
> (we don't need 6 "polishing" commits and fixups) and write meaningful messages for the rest.

There was no need pointing this, it's not my first pull request. It's strange to suggest I want it to merged in that state ;)

SuperTux88 commented:
> t's strange to suggest I want it to merged in that state ;)

I was confused because of this: "Ok, this feels completed to me.", sorry about that.

cmrd-senya commented:
> I was confused because of this: "Ok, this feels completed to me.", sorry about that.

No worries. Well, I didn't squash because there will be some new fixes after review anyway %) (you already discovered some issues)  But probably I had too. Next time I'll be armed with --autosquash! :D

allschild commented:
Sent from my BlackBerry 10 smartphone.                                                                                                                                                                                                                From: SenyaSent: Sunday 7 August 2016 01:06To: diaspora/diasporaReply To: diaspora/diasporaSubject: Re: [diaspora/diaspora] Mentions in comments backend changes (#6818)
(we don't need 6 "polishing" commits and fixups) and write meaningful messages for the rest.


There was no need pointing this, it's not my first pull request. It's strange to suggest I want it to merged in that state ;)

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/diaspora/diaspora","title":"diaspora/diaspora","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/modules/aws/aws-bg.jpg","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/diaspora/diaspora"}},"updates":{"snippets":[{"icon":"PERSON","message":"@cmrd-senya in #6818: \u003e (we don't need 6 \"polishing\" commits and fixups) and write meaningful messages for the rest.\r\n\r\nThere was no need pointing this, it's not my first pull request. It's strange to suggest I want it to merged in that state ;)"}],"action":{"name":"View Pull Request","url":"https://github.com/diaspora/diaspora/pull/6818#issuecomment-238044823"}}}

SuperTux88 commented:
Just found something more: If someone disables "you are mentioned in a post/comment" in the email-settings, but has enabled the "someone comments on ...", then he will not receive an email for this comment (because no commented-notification is created, and the mentioned notification is not sent).

I don't know, who should have this strange combination of settings, and I also don't know how to solve this the best, or if we just should ignore it. But I wanted to mention that here.

SuperTux88 commented on db/migrate/20160621151100_polymorphic_mentions.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def up_down(change)
+    change.up do
+      Mention.update_all(mentions_container_type: "Post")
```
Can you add [migration models](http://blog.makandra.com/2010/03/how-to-use-models-in-your-migrations-without-killing-kittens/) for the migrations?

Commit:
Mentions in comments backend changes

cmrd-senya commented:
> I don't know, who should have this strange combination of settings, and I also don't know how to solve this the best, or if we just should ignore it. But I wanted to mention that here.

I thought of a following possible solution: overload `email_the_user` for `MentionInComment` with a version that does something like

```ruby
class Notifications::MentionedInComment < Notification
  def email_the_user(target, actor)
    unless recipient.user_preferences.exists?(email_type: "mention"))
      recipient.mail(Workers::Mail::Mentioned, recipient_id, actor.id, target.id)
    elsif recipient.is_owner_of?(target)
      recipient.mail(Workers::Mail::CommentOnPost, recipient_id, actor.id, target.id)
    elsif recipient.participates_in?(target)
      recipient.mail(Workers::Mail::AlsoCommented, recipient_id, actor.id, target.id)
    end
  end
end
```

It's a bit hacky, since it repeats the logic of NotificationService and the other comment notifications. But it's the easiest thing I invented that doesn't require any refactoring of the notification code.

I also think it worth fixing, because I may imaging someone turns off "mention in comments" email notification, because he doesn't want to be spammed by people who want to include him into discussions. And after that someone mentions the person in a comment within a discussion he participated, and the person doesn't receive the email notification (but he expects it, because he is participating and doesn't care on whether someone mentions him or not). In this case the software will implicitly make other things than expected, which is not good.

Opinions?

cmrd-senya commented:
To simplify the review I've outlined the changes by this PR in the top message of the thread.

SuperTux88 commented on app/helpers/notifications_helper.rb:
```diff
@@ -4,40 +4,46 @@ module NotificationsHelper
 
+  def opts_for_post(post)
+    {
+      post_author: h(post.author_name),
```
Can you use `html_escape()` here? I needed my IDE to find what `h()` does.

SuperTux88 commented on app/helpers/notifications_helper.rb:
```diff
@@ -4,40 +4,46 @@ module NotificationsHelper
+      if note.linked_object.nil? && note.respond_to?(:deleted_translation_key)
+        target_type = note.deleted_translation_key
+      elsif note.is_a?(Notifications::Mentioned)
+        opts.merge!(opts_for_mentioned(note))
```
Maybe use `note.linked_object` here like for `opts_for_post`?

SuperTux88 commented on app/views/notifier/mentioned.markerb:
```diff
@@ -4,6 +4,12 @@
 <% end %>
 
-[<%= t('notifier.comment_on_post.reply', :name => @notification.post_author_name) %>][1]
+<% if @notification.post.is_a?(Post) %>
```
I think we should use different templates here. I don't like this if/else here, and now if you are mentioned in a private comment you get a mail with "You were mentioned in a limited post.".

SuperTux88 commented on app/models/mention.rb:
```diff
@@ -3,11 +3,16 @@
 class Mention < ActiveRecord::Base
-  belongs_to :post
+  belongs_to :mentions_container, polymorphic: :true
+  alias_attribute :post, :mentions_container
```
Do we need this alias here?

SuperTux88 commented on app/models/notifications/also_commented.rb:
```diff
@@ -18,7 +18,11 @@ def self.notify(comment, _recipient_user_ids)
       User.where(id: recipient_ids).find_each do |recipient|
-        next if recipient.is_shareable_hidden?(commentable)
+        if recipient.is_shareable_hidden?(commentable) ||
+          Notifications::MentionedInComment.where(target: comment.mentions.where(person: recipient.person)).exists?
```
You can extract `Notifications::MentionedInComment.where(target: comment.mentions.where(person: recipient.person)).exists?` to a method `mention_notification_exists?(comment, recipient)`. That way you can make this a one liner again and I think that would increase readability here.

cmrd-senya replied:
`mention_notification_exists?` is going to be used both in this and `CommentOnPost` classes. Is it fine to place this method to the `Notification` model? I'd place it to `Notifications::MentionedInComment` as class-method, however it wouldn't be one-liner then (I checked, it doesn't fit).

SuperTux88 replied:
Maybe you can also create a `Commented` module like the `Mentioned` with common logic, because `deleted_translation_key` is also the same for both. Otherwise I have no problem if it isn't a one-liner, I still think the method-name `mention_notification_exists?` is easier to read/understand than the AR-query and that was my main goal here. But if it is only used for commented-notifications it doesn't belong to `Notification`.

SuperTux88 commented on app/models/status_message.rb:
```diff
@@ -120,6 +89,29 @@ def post_location
+        "all"
+      else
+        [
+          *AspectMembership.where(aspect_id: aspects).map(&:person),
```
This line creates a DB-query for each person in this aspects, but this can be done with an all-in-one query:
```ruby
Contact.joins(:aspect_memberships).where(aspect_memberships: {aspect: aspects}).distinct.pluck(:person_id)
```

SuperTux88 commented on app/models/status_message.rb:
```diff
@@ -120,6 +89,29 @@ def post_location
+      else
+        [
+          *AspectMembership.where(aspect_id: aspects).map(&:person),
+          *share_visibilities.map {|sv| sv.user.person }
```
This line is not needed, because `people_allowed_to_be_mentioned` is only used with the author of the post, and the people from the aspects (see above) are all you need here.

SuperTux88 replied:
Oh, I missed the use in the `Notifications::Mentioned` ...

But I think it is better to create its own filter there, and only test if the person can see the post there (and for comments if the person can see the parent post)?

SuperTux88 commented on app/services/post_service.rb:
```diff
@@ -59,7 +59,11 @@ def mark_comment_reshare_like_notifications_read(post_id)
   end
 
   def mark_mention_notifications_read(post_id)
```
I think we should also mark comment-mention-notifications as read here.

SuperTux88 commented on spec/integration/mentioning_spec.rb:
```diff
@@ -22,74 +35,275 @@ def mention_stream_for(user)
 
-    @user1.share_with(@user2.person, default_aspect)
-    sign_in @user1
+    expect(users_connected?(alice, bob)).to be true
```
I think you should simply create what you need here (like before) instead of checking if nobody changes alice, bob and eve.

cmrd-senya replied:
What do you think of removing these checks at all and relying on the
fixture builder generated data?

SuperTux88 replied:
No, don't use alice, bob and eve from fixture builder and create user1, 2 and 3 like before and connect them like you need it.

SuperTux88 commented on spec/integration/mentioning_spec.rb:
```diff
@@ -22,74 +35,275 @@ def mention_stream_for(user)
+      let(:status_msg) { FactoryGirl.create(:status_message, author: alice.person, public: true) }
+
+      context "a commenter may mention a person" do
+        [[bob, eve], [eve, bob], [alice, bob], [bob, alice], [local_luke, local_leia], [alice, eve]]
```
Do we really need all of these here?

cmrd-senya replied:
Yes, these all are different type of relationships between users. I'll
check once again, but we need all of them because receive works
different depending on the relationship type. So it's easy to screw it up.

SuperTux88 replied:
Then you should probably create tests for the different combinations, so you can see which combination broke if one of these fails.

cmrd-senya replied:
What do you mean by "create tests for the different combinations"? I can
make context names express it more explicit, but the test body is the
same every time. Is it fine?

SuperTux88 replied:
Yes, same test, but write what "bob+eve" makes them special compared to "bob+alice".

SuperTux88 commented on spec/services/notification_service_spec.rb:
```diff
@@ -0,0 +1,56 @@
+            expect(Workers::Mail::Mentioned).to receive(:perform_async).with(
+              alice.id,
+              bob.person.id,
+              *comment.mentions.where(person: alice.person).ids
```
You can use `user` here to, so you don't test twice if alice receives a mail.

SuperTux88 commented on spec/services/notification_service_spec.rb:
```diff
@@ -0,0 +1,56 @@
+
+            expect(
+              Notifications::MentionedInComment.where(target: comment.mentions, recipient_id: user.id)
+            ).not_to be_empty
```
I think `exist` is better than `be_empty` here.

SuperTux88 commented:
> I thought of a following possible solution: overload `email_the_user` for `MentionInComment`

Hmm, I don't know. It fixes the problem and it is relatively simple. And I don't have a better idea, but we can always change that if someone has a better solution.

SuperTux88 commented on app/mailers/notification_mailers/mentioned.rb:
```diff
@@ -4,9 +4,10 @@ class Mentioned < NotificationMailers::Base
+      @post = Mention.find_by_id(target_id).mentions_container
 
       @headers[:subject] = I18n.t('notifier.mentioned.subject', :name => @sender.name)
+      @headers[:template_name] = "mentioned_in_comment" if post.is_a?(Comment)
```
I see what you are trying to do here ... and I'm unsure about it.

On one side you try to minimize code duplication, on the other side you need a workaround to make this work. Also you have `@post` here, which leads to stuff like `.post.parent` and `.post.guid` in the `mentioned_in_comment.markerb`.

If you create a separate class for `MentionedInComment` you would have a small code duplication, but you wouldn't need the workaround and you can name the variables properly (and you don't even need the `post_author_name` in the `mentioned_in_comment` template, so less duplication) and you don't need a `if` here.

And creating a separate class would also mean that you have separate checkboxes in the settings to enable mails for mentions on posts and comments.

At the end I'm more on the "split this completely"-side, I think that's the cleaner solution. What do you think?

cmrd-senya replied:
I'm fine with both approaches. If you like "split" approach more, I'm fine with implementing it.

However, speaking about deduplication, I see that the workers for Notifications::*  in `app/workers/mail` are almost identical. So are respective methods in `Notifier`. I can refactor them to use some common code instead of deepen the duplication by introducing `Workers::Mail::MentionedInComment`. What do you think?

SuperTux88 replied:
Creating a `Workers::Mail::Base` sounds like a good idea.

cmrd-senya replied:
However I'm not sure about `Workers::Mail::Base` as a good name. First it'll make confusion with `Workers::Base`. Second not every `Workers::Mail::*` will be child of this `Workers::Mail::Base`. There are `ReportWorker` or `InviteEmail` who have different logic. All `Workers::Mail::*` that we can commonify use `Notifier` class. So we can call the base class for it `Workers::Mail::NotifierBase` or something. ?

SuperTux88 replied:
Yes, `Workers::Mail::NotifierBase` is ok.

Commit:
Refactor mail workers to use common base

Introduce Workers::Mail::NotifierBase to be a base for all appropriate
mail workers to reduce code duplication

SuperTux88 commented on app/models/notifications/mentioned_in_post.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+
+    def self.people_allowed_to_be_notified(mentions_container)
+      super.tap do |ppl|
```
hmm, after thinking about this again: you don't even need this here, because `filter_mentions` does already the correct thing for mentions on posts. And for mentions on comments it is probably better to filter allowed mentions in `filter_mentions` too, so we don't have to filter twice.

SuperTux88 commented on app/models/comment.rb:
```diff
@@ -49,14 +51,24 @@ class Comment < ActiveRecord::Base
   end
 
+  def people_allowed_to_be_mentioned
+    @people_allowed_to_be_mentioned ||= if parent.public?
```
If I see it correctly, you only call this method once, so no need to cache the result in `@people_allowed_to_be_mentioned`?

cmrd-senya replied:
It is called twice on each comment generation, but in different workers, so caching doesn't help.

SuperTux88 commented on app/models/notifications/mentioned_in_comment.rb:
```diff
@@ -10,8 +10,15 @@ def deleted_translation_key
+      if ppl == "all"
+        mentions
+      else
+        mentions.to_ary.delete_if {|mention|
```
You can write this as `mentions.where(person_id: ppl)`.

cmrd-senya replied:
Right, but won't that be an inefficient `WHERE people.id IN (....)` expression? I remember before I was adviced to avoid this idiom in case when there are many ids in the list. Often there will be few mentions and a number of people ids (10, 100, even more).

SuperTux88 replied:
Yes, I was also unsure about this, but I talked with @CSammy about that, and he meant that the DB should manage it, because there are other `WHERE` conditions to minimize the result set. 

I did a few tests locally now, and yes, it gets slower if you add more IDs to the `IN`, but 10000 are still ~10ms, so it should be "ok" (not really slow). Also I don't expect it to be more than 100, because I only know public posts with really many participants, private posts are mostly less active (at least the private posts I know about).

But I'm still unsure ... I thought the `where` would be more readable, but if you prefer the `delete_if` I'm also fine with it. But maybe `keep_if` is more intuitive then?

cmrd-senya replied:
Well, the default scope is "All aspects" which means all contacts, but if it is fine with 10000, then I think it can hardly cause any issues.

I'd vote for readability and `where` clause then. If it cause any performance problems in future, it can be solved as a separate issue. I guess it comforms the Ruby/Rails way "features first, performance second".

SuperTux88 replied:
> "features first, performance second"

That shouldn't be an excuse to ignore performance ;)

But I think in this case the `where` is ok for the beginning, it shouldn't cause any issues, and if it does, we can change this in the future. This is also only in background processing (sidekiq), so no slow frontend actions because of this. But with 10000 it isn't fast anymore, compared to 0.x ms with 10, but I think we can live with it, it doesn't kill the DB.

> Well, the default scope is "All aspects"

That still doesn't mean that all contacts are participating.

SuperTux88 commented on app/models/comment.rb:
```diff
@@ -56,11 +56,11 @@ def text= text
+    if parent.public?
+      "all"
+    else
+      [*parent.participants, parent.author].map(&:id).uniq
```
you can do this with `[*parent.participations.pluck(:author_id), parent.author_id].uniq`, without joining the people table.

SuperTux88 replied:
While thinking about the filter in the notifications, I remembered, that this doesn't work as expected. With the new federation, `participations` only contain participations of all participants, if the owner of the post is local, otherwise it only contains the local participations.

While this is perfect for the notification filter (means less person_ids in the `IN` query, because it only contains local IDs), this doesn't work for the send filter. Do we really want all participants to be mentionable (comment, like or poll_participation?), or only people who commented?

I suggest the latter, so it would be: `[*parent.comments.pluck(:author_id), parent.author_id].uniq`

cmrd-senya replied:
Hmm, participations and poll_participations are generally non-visible. Comments and likes are visible. So there are few sense to include the former: strictly saying participation means "following" the discussion rather than participating in it and nobody sees your participation. On contrary comments are visible participation in the discussion. Likes are similar to comments in this respect. E.g. a user notices someone has liked the post and wants to say `@dude why do you like it?`

So, I guess, if changing that, I'd prefer 
`[*parent.comments.pluck(:author_id), *parent.likes.pluck(:author_id), parent.author_id].uniq`

Right?

SuperTux88 replied:
sounds good. 👍

cmrd-senya commented:
Ok, looks like I made all changes that was requested by @SuperTux88 so far. I'd love to test all of the changes for some more time, but I'm unsure I'll have the possibility to do so in the next days. It _should_ be fine, but I can't claim that I've conducted comprehensive examination.

cmrd-senya commented:
Okay, I've done some manual testing. I found a couple of minor issues and fixed them. If you think of some edge cases that I could have missed, please tell me, I can test them. Otherwise, everything works as expected for me and I think can be merged (after squash).

SuperTux88 reviewed (changes requested):
I found some things, but it looks good so far.

But I have not done much manual testing yet.

svbergerem commented on app/helpers/notifications_helper.rb:
```diff
@@ -4,40 +4,44 @@ module NotificationsHelper
   def object_link(note, actors)
     target_type = note.popup_translation_key
-    actors_count = note.actors.size
+    opts = {actors: actors, count: note.actors.size}
```
I guess @SuperTux88 was talking about the value instead of the key.

cmrd-senya replied:
That is what I wanted to clarify.

SuperTux88 replied:
Oh yeah, changing the locales is not what I intended. Your current change is OK, thanks 👍

SuperTux88 commented on spec/integration/mentioning_spec.rb:
```diff
@@ -19,77 +28,337 @@ def mention_stream_for(user)
+      let(:recipients) { [user2, user3] }
+    end
+
+    ["in the first possible recipient order", "in the second possible recipient order"].each do |in_context|
```
I don't know if we really need to test different orders here, but why not ... But I think we should probably use more meaningful names like "with user2 receiving before user3" or something like this.

cmrd-senya replied:
It was my original intention to test different orders with this examples, but I lost it somewhere on the way. When you commented on this part, I recalled it. Why I wanted this is because we receive a federation entity one by one and `Notifications::Mentioned.notify` is called with different recipient_user_ids. It is important to test that we don't fail if `Notifications::Mentioned.notify` was already called before with a different user id. For instance, a possible wrong flow is to receive entity for user2, filter out mention for user3 because there is no user3 among recipient_user_ids and then fail to notify user3 when we receive the same entity for the second time. That doesn't happen now, but who knows, how it can be refactored in future.

(sorry for double posting, github sent my message to the general PR discussion instead of this subthread)

SuperTux88 replied:
But does the order matter then?

Maybe you can also add some more unit-tests for `Notifications::Mentioned.notify`?

cmrd-senya replied:
> But does the order matter then?

On the second thought I think we don't really need this ordering test. I'll change it back to the fixed order.

> Maybe you can also add some more unit-tests for Notifications::Mentioned.notify?

Sure, why not.

cmrd-senya commented:
It was my original intention to test different orders with this
examples, but I lost it somewhere on the way. When you commented on this
part, I recalled it. Why I wanted this is because we receive a
federation entity one by one and Notifications::Mentioned.notify is
called with different recipient_user_ids. It is important to test that
we don't fail if Notifications::Mentioned.notify was already called
before with a different user id. For instance, a possible wrong flow is
to receive entity for user2, filter out mention for user3 because there
is no user3 among recipient_user_ids and then fail to notify user3 when
we receive the same entity for the second time. That doesn't happen now,
but who knows, how it can be refactored in future.

cmrd-senya commented:
I've addressed all the notes of @SuperTux88 so far.

cmrd-senya commented:
Ok, I made a self-review and fixed some issues. Please move the PR to "Needs review". Hopefully it'll be a final one. After that I'll squash it. Dunno if we need any more deep testing. Nobody replied on my call for testers at diaspora, however @dadall and @AugierLe42e discussed a possibility to merge it on diaspote. If we need some more testing before merge it is an opportunity.

AugierLe42e commented:
Yep, I'll merge this during this week-end. Is it already usable? Is the front-end present here?

Flaburgan commented:
No there is no front-end at the moment. I'm also open to merge this on d-fr if needed, though I agree with @SuperTux88 that it doesn't bring anything more than the local testing as it brings nothing new in the UI>

AugierLe42e commented:
I think we can write a quick UI then. The code is already available. We can reuse the mention system used in the publisher. I won't be a big deal.

cmrd-senya commented:
> I think we can write a quick UI then.

It's not necessary. We can test mentions writing mention markdown clause in comments directly: `@{name; id@pod.tld}`.  If we have a couple of pods running this PR, we can test any imaginable usecase live. Then, when this is merged, we can make the UI without rush.

AugierLe42e commented:
I'm not sure our users will use it with the `@{name;handle}` syntax. I will merge this in diaspote and add the UI. Does this PR have a path to query the users that can be mentionned? Like [there is for the publisher](https://github.com/diaspora/diaspora/blob/develop/app/controllers/contacts_controller.rb#L18)?

cmrd-senya commented:
> Does this PR have a path to query the users that can be mentionned?

No, there isn't since I haven't ever started the UI part.

> I'm not sure our users will use it with the @{name;handle} syntax.

The syntax is `@{name; handle}`, the space after `;` is important (that is a hardcoded regexp we have there). And I don't think it is necessary that every user of your pod will test this feature. It is fine to test it among a little group of people, even only ourselves. Again, testing it on "usual users" massively is not how I think we should do. It's better if manual testing is conducted by a person who have some technical skills and understanding of technologies. Most probably, that if a user fails to use mention markdown syntax correctly, then he would also fail to give a decent feedback. Then, if we rush with UI, it could make things more complicated, because you'll have to introduce even more changes, some of them to controllers. If there are any issues with the quickly made UI, it'll get harder to find what is broken.

It's enough even if one person (except of me) conducts testing if it is done well. If it is fine to test changes on the end-users right away, then why don't we just merge the PR to the develop branch right away without any more manual tests?

SuperTux88 commented:
> It's enough even if one person (except of me) conducts testing if it is done well.

Yes. I don't think that we have much benefit if we deploy this to a prod-pod and ask the users to test it. It is better when one person does a careful local testing with different users and as many edge cases the person can think of. This person can then also check the logs and check the expected behavior on every involved user. If you let multiple users test this in production, then they probably test it with their best friends (and always have a mutual connection), and that doesn't help much. But I can't stop somebody if he wants to test it in production, but I think local testing is easier and better.

AugierLe42e commented:
>  And I don't think it is necessary that every user of your pod will test this feature. It is fine to test it among a little group of people, even only ourselves. 

The users of our pod is already a small group. We are talking about less than 50 persons here ;)

> Then, if we rush with UI, it could make things more complicated, because you'll have to introduce even more changes, some of them to controllers.

Really, the UI is a one-liner. [mention_view.js](https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/app/views/publisher/mention_view.js) can be used as is, as long as there's a path to query mentionnable people. See, I'm not kidding, [it's really a one-liner](https://github.com/diaspora/diaspora/blob/develop/app/assets/javascripts/app/views/publisher_view.js#L99):

```javascript
this.mention = new app.views.PublisherMention({ el: this.$("#publisher_textarea_wrapper") });
```

That's it.

SuperTux88 commented on app/models/comment.rb:
```diff
@@ -49,14 +50,26 @@ class Comment < ActiveRecord::Base
+  def subscribers
+    super.tap {|subscribers|
+      if parent.public? && parent.author.local?
+        subscribers.concat(legitimate_mention_recipients.select(&:remote?))
```
This is now somehow redundant. You check `public?` here and also check public in `legitimate_mention_recipients` (which calls people_allowed_to_be_mentioned). Also we never need to add subscribers for private posts (status messages and comments)

I would rename `legitimate_mention_recipients` to `add_mention_recipients`, then you can do something like this:
```ruby
    def add_mention_recipients(subscribers)
      subscribers.concat(mentions.map(&:person).select(&:remote?)) if public?
    end
```

And you can use this for comments and status messages.

Just some quick thoughts, maybe you have a better idea how to simplify this? :)

cmrd-senya replied:
see https://github.com/diaspora/diaspora/pull/6818/commits/39cd3caf5f33fc1f7b98f73171675bc465afc19a

SuperTux88 commented on spec/models/notifications/mentioned_spec.rb:
```diff
@@ -1,72 +1,58 @@
+      expect(TestNotification).to receive(:filter_mentions).with(
+        Mention.where(mentions_container: status_message, person: [alice, bob].map(&:person)),
+        status_message,
+        :passthrough_argument
```
This confused me, what about `[alice.id, bob.id]`?

SuperTux88 commented on spec/models/notifications/mentioned_spec.rb:
```diff
@@ -1,72 +1,58 @@
 
-        Notifications::Mentioned.notify(private_sm, [alice.id])
-      end
+    it "doesnt't create notification if it was filtered out by filter_mentions" do
```
"doesn't ..."

SuperTux88 commented on spec/models/notifications/mentioned_spec.rb:
```diff
@@ -1,72 +1,58 @@
-      expect(bob).to receive(:mail).with(Workers::Mail::Mentioned, bob.id, sm.author.id, sm.mentions.first.id)
-
-      Notifications::Mentioned.notify(sm, [])
+    def email_the_user(*)
```
You can remove this.

SuperTux88 reviewed (changes requested):
Temporary review (but I wanted to give some quick feedback, so you can already fix some things, so we can merge this faster :) ).

I'll do a more careful review again on the weekend and also some testing.

SuperTux88 commented on app/models/status_message.rb:
```diff
@@ -111,10 +111,8 @@ def filter_mentions
-      subscribers.concat(legitimate_mention_recipients)
-    }
+  def add_mention_subscribers?
+    super && author.local?
```
Do we really need the `author.local?` here? When sending a StatusMessage the author is always local, and the subscribers of the StatusMessage are only added to the relayable-subscribers if the author is local.

SuperTux88 commented on features/desktop/notifications.feature:
```diff
@@ -101,6 +101,26 @@ Feature: Notifications
+    Then I should see "mentioned you in the comment"
+    And I should have 1 email delivery
+
+  Scenario: I mark a mention in comment notification as read
```
> Why name it generic "mark a notification as read" if it tests mark as read for "mention in comment" notification?

Because it doesn't matter. The "mark as read/unread" feature doesn't depend on the notification type. So it is OK to have a scenario that tests it in the frontend, but I don't want cucumber tests for every notification type to test if you can mark it as read. (that is something you can do in unit-test, which are faster).

And if it doesn't depend on the notification type, I think we should use a generic description. And because you don't need "mention in comment" you can also test it with a normal "mention" and use the existing "And Alice has a post mentioning Bob" step for it.

cmrd-senya replied:
> Because it doesn't matter. The "mark as read/unread" feature doesn't depend on the notification type. So it is OK to have a scenario that tests it in the frontend, but I don't want cucumber tests for every notification type to test if you can mark it as read. (that is something you can do in unit-test, which are faster).

I disagree. IMO, ideally, we should have an integration tests covering each client behaviour scenario. We can't rely on unit tests here. Unit tests are not for prooving that our application works as expected in every usage scenario. Ideally, we should have an integration test for every possible situation that may happen from the user point of view. The execution flow for "mark as read" for mentions in comments and posts is the same, but that is an implementation detail. We can't use that knowledge to claim that 
> "mark as read/unread" feature doesn't depend on the notification type.

In the end, I'll implement what you say, but I don't feel it right to have only one test scenario for "mark as read" for mentioning.

svbergerem replied:
I agree with @SuperTux88 here. We use cukes to test the combination of front- and backend. And when doing that at least I also have the implementation in mind to know, which cases I have to go through.

You could use different “mark read/unread” cukes for different notification types. You could also use different “mark read/unread” cukes for contact / non-contact actors. Or users on the same pod vs. users on different pods. Or for posts with photos and posts without photos. Or... (I hope you get my point ;))

In this case there is exactly one thing we have to test: Do we send the correct request to the correct route when clicking the button. And the requests all look the same for the different notification types. And the buttons also look the same.

I still remember the times when our cucumber tests exceeded the timeout of 50 minutes. Now they need about 15 minutes but we have a few problems with randomly failing tests. Thus I'd still like to keep the number of cukes rather small. Let's keep using cukes to test the different combinations of front- and backend. And let's keep using rspec/jasmine tests for everything else. In this case I don't think that we need more than one cuke for the “mark read/unread” feature.

AugierLe42e commented:
Keep up the good work @cmrd-senya!

![img-img-1](https://cloud.githubusercontent.com/assets/3429231/20635165/a51a0078-b359-11e6-8ae8-5bc0be6b8351.gif)

SuperTux88 reviewed (changes requested):
Looks really good so far, I think we can merge this soon :)

I also did some local testing now with different scenarios, and everything works as expected 👍 

Can you please fix these last remarks and then squash your commits? :)

SuperTux88 commented:
I have just merged #7223 and that causes some conflicts with this PR (sorry about that, but they should be easy to resolve :) ). When rebasing, can you please also remove `require "spec_helper"` from any new specs you added (where you don't have a conflict).

cmrd-senya commented:
Remarks corrected, squashed, rebased, signed. Tests are passing, except of random cukes.

SuperTux88 commented:
Merged to develop branch, thanks @cmrd-senya :)

cmrd-senya commented:
Great! Thanks for your work either. I know it required many much efforts to review it multiple times. I'll do my best to produce a better code in my next contributions :)

We also need UI to get #1851 fixed. I'm going to work on it soon. I don't really think it is a one-liner, as @AugierLe42e says, when I looked at it last time I didn't quite get it. But that's on my list anyway.

goobertron commented:
Whoo whoo whoo! Thanks so much, @cmrd-senya, and thanks to all those who helped and reviewed.

Flaburgan commented:
A massive thank you for this very needed work, this was definitely a missing key feature on diaspora.

AugierLe42e commented:
Same as @Flaburgan : huge thanks to @cmrd-senya and @SuperTux88.

> when I looked at it last time I didn't quite get it. But that's on my list anyway.

As explained, as soon as there's a route that answers mentionnable people for a particular post, it's easy. Maybe not a one-liner, but 3 or 4 lines at most. As I refactored the mention system, I can help you or even write it myself. Just ping me on IRC if you need it.
